plugins/analytics/public/packettotal.py

```diff
@@ -0,0 +1,158 @@
+import logging
+
+import requests
+
+from core.analytics import OneShotAnalytics
+from core.config.config import yeti_config
+from core.observables import Hostname, Ip, Text
+
+
+class PacketTotalAPI(object):
+    """Base class for querying the PacketTotal API."""
+    settings = {
+        'packettotal_api_key': {
+            'name': 'PacketTotal API Key',
+            'description': 'API Key provided by packettotal.com'
+        }
+    }
+
+    API = 'https://api.packettotal.com/v1/'
+
+    @staticmethod
+    def search(observable, api_key):
+        # """
+        # param observable: The extended observable class
+        # param api_key: The api key obtained from PacketTotal
+        # return  packettotal json response or None if error
+        # """
+
+        params = {'query': observable.value}
+
+        try:
+            res = requests.get(
+                '{}search'.format(PacketTotalAPI.API),
+                headers={'x-api-key': api_key},
+                params=params,
+                proxies=yeti_config.proxy
+            )
+
+            if res.ok:
+                return res.json()
+
+        except Exception as e:
+            logging.error(
+                'Exception while retreiving packettotal report {}'.
+                format(e.message))
+
+    @staticmethod
+    def fetch_analysis(pcap_id, api_key):
+        # """
+        # param observable: The extended observable class
+        # param api_key: The api key obtained from PacketTotal
+        # return:  packettotal json response or None if error
+        # """
+
+        try:
+            res = requests.get(
+                '{}pcaps/{}/analysis'.format(PacketTotalAPI.API, pcap_id),
+                headers={'x-api-key': api_key},
+                verify=False,
+                proxies=yeti_config.proxy
+            )
+
+            if res.ok:
+                return res.json()
+
+        except Exception as e:
+            logging.error(
+                'Exception while retreiving packettotal report {}'
+                .format(e.message))
+
+
+class PacketTotalQuery(PacketTotalAPI, OneShotAnalytics):
+    default_values = {
+        'name': 'PacketTotal',
+        'description': 'Perform a PacketTotal query.',
+    }
+
+    ACTS_ON = ['Ip', 'Hostname', 'Hash', 'Url']
+
+    @staticmethod
+    def analyze(observable, results):
+        links = set()
+        pcap_hits = PacketTotalAPI.search(
+            observable, results.settings['packettotal_api_key'])
+
+        result = {}
+
+        if pcap_hits:
```

Comment:
Tip: you can save on indentation issues in python by using guard statements.

If you replace 

```python
if pcap_hits:
   # blah
```

by

```python
if not pcap_hits:
    return []
# blah
```

You'll save one indentation level for the rest of your logic (`# blah`)

You can do this here as well as L95, saving another indentation level.

Reply:
Thanks for the tip :+1: